# Welcome to lab_🐼s!

Last week, you did amazing work setting up your computer to do data science!  Let's now use your computer to do data science for the first time!

A few tips to remember:

- **You are not alone on your journey in learning programming!**  You have your lab Teaching Assistant (TA), your Course Assistants (CAs), your lab group, and the professors (Prof. Karle and Prof. Wade), who are all here to help you out!
- If you find yourself stuck for more than a few minutes, ask a neighbor or course staff for help!  When you are giving help to your neighbor, explain the **idea and approach** to the problem without sharing the answer itself so they can have the same **<i>ah-hah</i>** moment!
- We are here to help you!  Don't feel embarrassed or shy to ask us for help!

Let's get started!

<hr style="color: #DD3403;">

# Part 0: Your Lab Group!

We find it's best when you're programming with a group of people who you can help out and can help you out!

- A key part of Data Science is being able to **communicate about your work**.  If you are the best programmer in the world, but no one uses your code, then the code you're writing is not getting run.  By learning strong communication skills, you will be able to **explain your code**, and the underlying data science concepts, to everyone -- data scientists and non-scientists alike!

- **Every lab** you will need to complete some parts of the lab **with your group**.  There will be both **"group discussions"** and **"individual reflections"** as part of every lab.

## Create Your Group

If you have not done so already, **RIGHT NOW** find *(ideally)* two other people to make a group of **two, three, or four** people.   *(We find groups of three work the best, but you can be a group of two or four -- no solo groups, and no groups of five!*)

- If you're at a group of five, split into a group of two and a group of three.  You can still help each other as a larger group, but the group questions should be answered in your smaller group of 2-4 people.

Once you have your group, **introduce yourselves to each other! :)**

## Record Your Group

Since you know Python, edit the following cell to add strings about your group:

In [30]:
# First, meet your CAs and TA if you haven't already!
# ...first name is enough, we'll know who they are! :)
ta_name = "Jake"
ca1_name = "Mia"
ca2_name = "Jessica"
ca3_name = ""


# As part of your introduction, learn your lab partner's name, major, and
# learn something new about them that you never knew!
# - Groups of 3 are ideal :)
# - However groups of 2 or 4 are fine too!
partner1_name = "Haibo Che"
partner1_netid = "haiboc2"
partner1_major = "econometrics"

partner2_name = "Hongyu Qi"
partner2_netid = "hongyuq2"
partner2_major = "econometrics"

partner3_name = ""
partner3_netid = ""
partner3_major = ""

<hr style="color: #DD3403;">

# Part 1: 🧑‍🤝‍🧑 Hello Dataset!

In lecture, we have already begun to explore the "Party Dataset" that you created.  We will continue to use the Party Dataset all semester long, but we want even more datasets -- and we want you to come up with the questions!

In this lab, **you** will come up with questions to ask the entire class for the **"Hello Dataset"**!

## Hello Dataset Overview

Over the weekend, your lab homework will be to fill out a class-wide "Hello Survey".  Right now, with your group, you will come up some of the questions that should be on this survey *(and you will do analysis on your questions next week!)*.

With your group, come up with four questions below!  Before you get started, a few notes:

- The survey will be **NOT be anonymous** and the results will be shared with the class to do analysis on, so you should **only ask questions you feel comfortable to share with the class**.
- **Questions that are inappropriate** for a classroom environment will be thrown out and you won't earn lab credit points for any question we throw out.  (Keep everything at most PG/PG-13.)
- You are going to submit **one set of questions for your whole group**.  You must decide on the questions together.

## Q1: Question with a Binary Response

A question with a **binary response** is a question that has only two answers.  For example:
- *Do you prefer cats or dogs?*
- *Were you born in Illinois?*

With your group (you must all decide on a single question together), come up with a question that you think will have **exactly 50%/50%** split between the two binary outcomes.  *(The question that gets closest to 50%/50% will get extra credit!)*



*Would you rather lose access to all social media for a year or lose access to all streaming services for a year?*

**Why do you think this is a good question?  Write at least two complete sentences on why you believe your question will have a 50%/50% split.**


* The question only have two answers. The question do not have preference.*

## Q2: Question with a Numeric Response

A question with a **numeric response** requires a number for an answer.  For example:

- *What is your height?*
- *What day were you born?*

With your group (you must all decide on a single question together), come up with a question and a prediction on what you think the most frequent answer will be to your question.

**What is your numeric question?**

*( What is your weight? )*

**What do you think the most frequent answer will be?**

*(65)*

**Why do you think this answer will be the most frequent answer?  Write at least one complete sentence on why you believe that it will be the most frequent answer.**

*(The average of human weight is about 65kg,including male and female)*

## Q3: Question with a (non-numeric) Textual Response

A question with a (non-numeric) **textual response** requires a word or two for an answer.  For example:

- *What is your favorite sport to play?*
- *What is the location of your favorite trip/vacation?*

With your group (you must all decide on a single question together), come up with a question where at least two members of your group share the same answer.

**What is your textual response question?**

*(What is your favorite music)*

**What is the shared answer between your group?  Do you also share this answer (if not, what's your answer to the question)?**

*("Die for you" my answer is Black Swan by BTS)*

## Q4: Open Question

One final question -- no restrictions on the format!  Is there a great question your group discussed that you did not get to use as one of the first three questions but you want to know the class's answer?

**What is your final question for the whole class?**

*(Do you like STAT107?)*

## Submit Your Questions to Professors Wade and Karle

Your lab TAs will be grading your lab notebook for your full responses, but we want to make sure to get the survey with your questions out to everyone ASAP!

- Pick one member of your group as the "question submitter".
- If you're the question submitter, complete this Google Form for your group (requires @illinois.edu login): https://forms.gle/6HQ6MLyEhq7zK36i9
- Make sure to do this **during your lab section**.


## This Weekend: Complete the "Hello Survey"

On Friday evening, after the final lab section, we will post the "Hello Survey" with the questions you (and the entire class) asked.  You will have to complete this before **Monday, Sept. 9** and it will be part of your lab grade.

<hr style="color: #DD3403;">

# Part 2: Importing the Course Catalog Dataset

In lecture, you learned we will use the `pandas` **library** for working with data.  Libraries are collections of functions and other objects that extend what Python can do by itself.  The `pandas` library is the most widely used library for data science.

To use the `pandas` library, we must **import** it into your notebook.  In every lab, you will need to import the library just like we did in lecture!

In the following cell, replace `...` with the code to import pandas as `pd`:

In [31]:
# Write the code to import pandas as pd:
import pandas as pd

In lecture, you saw the "Course Catalog" dataset.  Using the `read_csv` function, load the course catalog dataset.  The URL for this dataset is:
```
https://waf.cs.illinois.edu/discovery/course-catalog.csv
```

Remember, URLs passed to `read_csv` are **string** variables, so remember to use **quotation marks**. 

> **macOS users**: If you are running macOS, you may get a HTTPS connection security-related error.  If you get this error, there's a solution for you on the Canvas lab overview page!! :)

In [32]:
# Load the course catalog dataset into a variable named `df`:
df = pd.read_csv("https://waf.cs.illinois.edu/discovery/course-catalog.csv")

In [33]:
# Look at `df` to confirm that this worked:
df

Year  Term YearTerm Subject  Number                          Name  \
0      2024  Fall  2024-fa     AAS     100  Intro Asian American Studies   
1      2024  Fall  2024-fa     AAS     100  Intro Asian American Studies   
2      2024  Fall  2024-fa     AAS     100  Intro Asian American Studies   
3      2024  Fall  2024-fa     AAS     100  Intro Asian American Studies   
4      2024  Fall  2024-fa     AAS     100  Intro Asian American Studies   
...     ...   ...      ...     ...     ...                           ...   
12985  2024  Fall  2024-fa    WRIT     303          Writing Across Media   
12986  2024  Fall  2024-fa    WRIT     303          Writing Across Media   
12987  2024  Fall  2024-fa    YDSH     101           Beginning Yiddish I   
12988  2024  Fall  2024-fa    YDSH     101           Beginning Yiddish I   
12989  2024  Fall  2024-fa    YDSH     220           Jewish Storytelling   

                                             Description Credit Hours  \
0      Interdisciplinary introduction to the basic co...     3 hours.   
1      Interdisciplinary introduction to the basic co...     3 hours.   
2      Interdisciplinary introduction to the basic co...     3 hours.   
3      Interdisciplinary introduction to the basic co...     3 hours.   
4      Interdisciplinary introduction to the basic co...     3 hours.   
...                                                  ...          ...   
12985                    Same as INFO 303. See INFO 303.     3 hours.   
12986                    Same as INFO 303. See INFO 303.     3 hours.   
12987  Course develops basic conversational and readi...     4 hours.   
12988  Course develops basic conversational and readi...     4 hours.   
12989  Course will introduce the great Jewish storyte...     3 hours.   

                                          Section Info  \
0                                                  NaN   
1                                                  NaN   
2                                                  NaN   
3                                                  NaN   
4                                                  NaN   
...                                                ...   
12985                  Same as INFO 303. See INFO 303.   
12986                  Same as INFO 303. See INFO 303.   
12987                                              NaN   
12988                                              NaN   
12989  Same as CWL 221, ENGL 223, JS 220, and REL 220.   

                                       Degree Attributes  ... Section Status  \
0      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
1      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
2      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
3      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
4      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
...                                                  ...  ...            ...   
12985                       Advanced Composition course.  ...              A   
12986                       Advanced Composition course.  ...              A   
12987                                                NaN  ...              A   
12988                                                NaN  ...              A   
12989  Humanities - Lit & Arts, and Cultural Studies ...  ...              A   

                Enrollment Status                   Type Type Code Start Time  \
0                            Open  Discussion/Recitation       DIS   09:00 AM   
1                            Open  Discussion/Recitation       DIS   10:00 AM   
2                            Open  Discussion/Recitation       DIS   10:00 AM   
3                            Open  Discussion/Recitation       DIS   11:00 AM   
4                            Open  Discussion/Recitation       DIS   11:00 AM   
...                           ...                    ...       ...       

## 🔬 Test Case Checkpoint 🔬

Throughout the lab, we will have "checkpoints" at key points.  Most checkpoints will include **test cases** that verify that your notebook appears to be correct.

- The test cases are **NOT** a guarantee that everything is perfect, but we try our very best to catch any common mistakes you might make!
- If you see `AssertionError` error message when running the test case, we provide a helpful hint on what may not be correct.
- If you see a party popper emoji and a message that all tests pass, you passed the checkpoint!
- **Checkpoints** are a good time to check with your group to make everyone gets to the checkpoint together!

In [34]:
## == CHECKPOINT TEST CASES for Part 1 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df' in vars() ), "Did you define a variable called df?"
assert( len(df) == 12990 ), "It looks like you did not import the correct dataset?"


## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


<hr style="color: #DD3403;">

# Part 3: STAT / CS / IS Majors

Data Science DISCOVERY is a unique course that is cross-listed across three departments: Statistics, Computer Science, and Information Sciences.  Let's explore courses from all three of those subjects!

### Puzzle 3.1

Select only the rows that have a `Subject` of `STAT`.  Assign this to the variable name `df_STAT` instead of the usual `df`.
- Make sure your result contains only statistics courses!

In [35]:
df_STAT = df[df["Subject"]=="STAT"]
df_STAT

Year  Term YearTerm Subject  Number                           Name  \
12180  2024  Fall  2024-fa    STAT     100                     Statistics   
12181  2024  Fall  2024-fa    STAT     100                     Statistics   
12182  2024  Fall  2024-fa    STAT     100                     Statistics   
12183  2024  Fall  2024-fa    STAT     107         Data Science Discovery   
12184  2024  Fall  2024-fa    STAT     107         Data Science Discovery   
...     ...   ...      ...     ...     ...                            ...   
12351  2024  Fall  2024-fa    STAT     571          Multivariate Analysis   
12352  2024  Fall  2024-fa    STAT     590  Individual Study and Research   
12353  2024  Fall  2024-fa    STAT     593                STAT Internship   
12354  2024  Fall  2024-fa    STAT     593                STAT Internship   
12355  2024  Fall  2024-fa    STAT     599                Thesis Research   

                                             Description    Credit Hours  \
12180  First course in probability and statistics at ...        3 hours.   
12181  First course in probability and statistics at ...        3 hours.   
12182  First course in probability and statistics at ...        3 hours.   
12183  Data Science Discovery is the intersection of ...        4 hours.   
12184  Data Science Discovery is the intersection of ...        4 hours.   
...                                                  ...             ...   
12351  Inference in multivariate statistical populati...        4 hours.   
12352  Directed reading and research. Approved for le...   0 TO 8 hours.   
12353  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
12354  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
12355  Approved for S/U grading only. May be repeated...  0 TO 16 hours.   

                                            Section Info  \
12180  Credit is not given for both STAT 100 and any ...   
12181  Credit is not given for both STAT 100 and any ...   
12182  Credit is not given for both STAT 100 and any ...   
12183                         Same as CS 107 and IS 107.   
12184                         Same as CS 107 and IS 107.   
...                                                  ...   
12351  Prerequisite: STAT 410 and MATH 415, or consen...   
12352  Approved for letter and S/U grading. May be re...   
12353  Approved for letter and S/U grading. Prerequis...   
12354  Approved for letter and S/U grading. Prerequis...   
12355  Approved for S/U grading only. May be repeated...   

                      Degree Attributes  ... Section Status  \
12180  Quantitative Reasoning I course.  ...              A   
12181  Quantitative Reasoning I course.  ...              A   
12182  Quantitative Reasoning I course.  ...              A   
12183  Quantitative Reasoning I course.  ...              A   
12184  Quantitative Reasoning I course.  ...              A   
...                                 ...  ...            ...   
12351                               NaN  ...              A   
12352                               NaN  ...              A   
12353                               NaN  ...              A   
12354                               NaN  ...              A   
12355                               NaN  ...              A   

       Enrollment Status                Type Type Code Start Time  End Time  \
12180             Closed  Lecture-Discussion       LCD   12:00 PM  12:50 PM   
12181  Open (Restricted)  Lecture-Discussion       LCD   03:00 PM  03:50 PM   
12182             Closed              Online       ONL   ARRANGED       NaN   
12183             Closed             Lecture       LEC   11:00 AM  11:50 AM   
12184             Closed             Lecture       LEC   09:00 AM  09:50 AM   
...                  ...                 ...       ...        ...       ...   
12351  Open (Restricted)  Lecture-Discussion       LCD   03:00 PM  03:50 PM   
12352  Open (Restricted)   Independent Study       IND   AR

### Puzzle 3.2

Select only the rows that have a `Subject` of `CS`.  Assign this to the variable name `df_CS` instead of the usual `df`.
- Make sure your result contains only computer science courses!

In [36]:
df_CS = df[df['Subject'] == 'CS']
df_CS

Year  Term YearTerm Subject  Number                              Name  \
3298  2024  Fall  2024-fa      CS     100      Computer Science Orientation   
3299  2024  Fall  2024-fa      CS     101  Intro Computing: Engrg &amp; Sci   
3300  2024  Fall  2024-fa      CS     101  Intro Computing: Engrg &amp; Sci   
3301  2024  Fall  2024-fa      CS     101  Intro Computing: Engrg &amp; Sci   
3302  2024  Fall  2024-fa      CS     101  Intro Computing: Engrg &amp; Sci   
...    ...   ...      ...     ...     ...                               ...   
3768  2024  Fall  2024-fa      CS     598                    Special Topics   
3769  2024  Fall  2024-fa      CS     598                    Special Topics   
3770  2024  Fall  2024-fa      CS     598                    Special Topics   
3771  2024  Fall  2024-fa      CS     598                    Special Topics   
3772  2024  Fall  2024-fa      CS     599                   Thesis Research   

                                            Description    Credit Hours  \
3298  Introduction to Computer Science as a field an...        1 hours.   
3299  Fundamental principles, concepts, and methods ...        3 hours.   
3300  Fundamental principles, concepts, and methods ...        3 hours.   
3301  Fundamental principles, concepts, and methods ...        3 hours.   
3302  Fundamental principles, concepts, and methods ...        3 hours.   
...                                                 ...             ...   
3768  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3769  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3770  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3771  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3772    Approved for S/U grading only. May be repeated.  0 TO 16 hours.   

                                           Section Info  \
3298                                                NaN   
3299  Prerequisite: One of MATH 220 or MATH 221 or M...   
3300  Prerequisite: One of MATH 220 or MATH 221 or M...   
3301  Prerequisite: One of MATH 220 or MATH 221 or M...   
3302  Prerequisite: One of MATH 220 or MATH 221 or M...   
...                                                 ...   
3768  May be repeated in the same or separate terms ...   
3769  May be repeated in the same or separate terms ...   
3770  May be repeated in the same or separate terms ...   
3771  May be repeated in the same or separate terms ...   
3772    Approved for S/U grading only. May be repeated.   

                      Degree Attributes  ... Section Status  \
3298                                NaN  ...              A   
3299  Quantitative Reasoning II course.  ...              A   
3300  Quantitative Reasoning II course.  ...              A   
3301  Quantitative Reasoning II course.  ...              A   
3302  Quantitative Reasoning II course.  ...              A   
...                                 ...  ...            ...   
3768                                NaN  ...              A   
3769                                NaN  ...              A   
3770                                NaN  ...              A   
3771                                NaN  ...              A   
3772                                NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
3298  Open (Restricted)                Lecture       LEC   03:00 PM  03:50 PM   
3299  Open (Restricted)     Lecture-Discussion       LCD   10:00 AM  10:50 AM   
3300               Open  Laboratory-Discussion       LBD   01:00 PM  02:50 PM   
3301               Open  Laboratory-Discussion       LBD   03:00 PM  04:50 PM   
3302               Open  Laboratory-Discussion       LBD   05:00 PM  06:50 PM   
...                 ...                    ...       ...        ...       ...   
3768  Open (Restricted)     Lecture-Discussion       S10   09:30 AM  10:45 AM   
3769             Closed     Lecture-Discus

### Puzzle 3.3

Select only the rows that have a `Subject` of `IS`.  Assign this to the variable name `df_IS` instead of the usual `df`.
- Make sure your result contains only information science courses!

In [37]:
df_IS = df[df['Subject'] == 'IS']
df_IS


Year  Term YearTerm Subject  Number  \
7003  2024  Fall  2024-fa      IS     100   
7004  2024  Fall  2024-fa      IS     100   
7005  2024  Fall  2024-fa      IS     100   
7006  2024  Fall  2024-fa      IS     100   
7007  2024  Fall  2024-fa      IS     100   
...    ...   ...      ...     ...     ...   
7303  2024  Fall  2024-fa      IS     599   
7304  2024  Fall  2024-fa      IS     599   
7305  2024  Fall  2024-fa      IS     599   
7306  2024  Fall  2024-fa      IS     599   
7307  2024  Fall  2024-fa      IS     599   

                                                  Name  \
7003  Exploring the iSchool with a Human-Centered Lens   
7004  Exploring the iSchool with a Human-Centered Lens   
7005  Exploring the iSchool with a Human-Centered Lens   
7006  Exploring the iSchool with a Human-Centered Lens   
7007  Exploring the iSchool with a Human-Centered Lens   
...                                                ...   
7303                                   Thesis Research   
7304                                   Thesis Research   
7305                                   Thesis Research   
7306                                   Thesis Research   
7307                                   Thesis Research   

                                            Description    Credit Hours  \
7003  This course introduces students to the School ...        1 hours.   
7004  This course introduces students to the School ...        1 hours.   
7005  This course introduces students to the School ...        1 hours.   
7006  This course introduces students to the School ...        1 hours.   
7007  This course introduces students to the School ...        1 hours.   
...                                                 ...             ...   
7303  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   
7304  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   
7305  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   
7306  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   
7307  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info Degree Attributes  \
7003  Prerequisite: Restricted to Majors Only; First...               NaN   
7004  Prerequisite: Restricted to Majors Only; First...               NaN   
7005  Prerequisite: Restricted to Majors Only; First...               NaN   
7006  Prerequisite: Restricted to Majors Only; First...               NaN   
7007  Prerequisite: Restricted to Majors Only; First...               NaN   
...                                                 ...               ...   
7303  0 to 16 graduate hours. No professional credit...               NaN   
7304  0 to 16 graduate hours. No professional credit...               NaN   
7305  0 to 16 graduate hours. No professional credit...               NaN   
7306  0 to 16 graduate hours. No professional credit...               NaN   
7307  0 to 16 graduate hours. No professional credit...               NaN   

      ... Section Status  Enrollment Status                Type Type Code  \
7003  ...              A             Closed  Lecture-Discussion       LCD   
7004  ...              A             Closed  Lecture-Discussion       LCD   
7005  ...              A             Closed  Lecture-Discussion       LCD   
7006  ...              A             Closed  Lecture-Discussion       LCD   
7007  ...              A             Closed  Lecture-Discussion       LCD   
...   ...            ...                ...                 ...       ...   
7303  ...              A               Open   Independent Study       IND   
7304  ...              A               Open   Independent Study       IND   
7305  ...              A               Open   Independent Study       IND   
7306  ...              A               Open   Independent Study       IND   
7307  ...              A               Open   Independent Study       IND   

     Start Time  

### 🔬 Test Case Checkpoint 🔬


In [38]:
## == CHECKPOINT TEST CASES for Part 3 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df_CS' in vars() ), "Make certain to name the CS courses df_CS."
assert( 'df_STAT' in vars() ), "Make certain to name the STAT courses df_STAT."
assert( 'df_IS' in vars() ), "Make certain to name the IS courses df_IS."
assert( len(df_STAT[df_STAT.Subject != "STAT"] ) == 0 ), "It looks like you did not subset df_STAT to only STAT courses."
assert( len(df_CS[df_CS.Subject != "CS"] ) == 0 ), "It looks like you did not subset df_CS to only CS courses."
assert( len(df_IS[df_IS.Subject != "IS"] ) == 0 ), "It looks like you did not subset df_IS to only IS courses."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Puzzle 3.4

Looking at the output of the DataFrames for STAT, CS, and IS above, which subject offers the **fewest** course sections?  The **most** course sections?

Be very clear on which specific subject has the **fewest** sections and which specific subject has the **most** sections by writing a complete sentence to answer this question.

*(STAT offers the fewest course sections, CS offers the most course sections)*


### Puzzle 3.5

Looking at the output of the DataFrames for STAT, CS, and IS above, which subject has the highest course number (for example: "STAT 107" course number is "107")?  Is there something similar about all of the highest numbered courses?  Write about what you observed!


*(All of the highest numbered courses are 599)*

<hr style="color: #DD3403;">

# Part 4: Your Major

In this part, you will be exploring **your major** using the Course Catalog dataset.

- Use the primary subject of your **current major** (or, if you're undeclared, your intended major) for this section.
- Your answers will be different than the rest of your group!

### Puzzle 4.1:

Select only the rows that have a `Subject` of your current/intended major.  Assign this to the variable name `mydf` instead of `df`:

In [39]:
mydf =  df[df['Subject'] == 'ECON']
mydf

Year  Term YearTerm Subject  Number                           Name  \
4514  2024  Fall  2024-fa    ECON     102       Microeconomic Principles   
4515  2024  Fall  2024-fa    ECON     102       Microeconomic Principles   
4516  2024  Fall  2024-fa    ECON     102       Microeconomic Principles   
4517  2024  Fall  2024-fa    ECON     102       Microeconomic Principles   
4518  2024  Fall  2024-fa    ECON     102       Microeconomic Principles   
...    ...   ...      ...     ...     ...                            ...   
4717  2024  Fall  2024-fa    ECON     590  Individual Study and Research   
4718  2024  Fall  2024-fa    ECON     598  Workshop and Research Seminar   
4719  2024  Fall  2024-fa    ECON     598  Workshop and Research Seminar   
4720  2024  Fall  2024-fa    ECON     598  Workshop and Research Seminar   
4721  2024  Fall  2024-fa    ECON     598  Workshop and Research Seminar   

                                            Description   Credit Hours  \
4514  Introduction to the functions of individual de...       3 hours.   
4515  Introduction to the functions of individual de...       3 hours.   
4516  Introduction to the functions of individual de...       3 hours.   
4517  Introduction to the functions of individual de...       3 hours.   
4518  Introduction to the functions of individual de...       3 hours.   
...                                                 ...            ...   
4717  Directed reading and research. Approved for bo...  0 TO 4 hours.   
4718  Workshops are offered in all areas of speciali...       2 hours.   
4719  Workshops are offered in all areas of speciali...       2 hours.   
4720  Workshops are offered in all areas of speciali...       2 hours.   
4721  Workshops are offered in all areas of speciali...       2 hours.   

                                           Section Info  \
4514      Credit is not given for ECON 102 and ACE 100.   
4515      Credit is not given for ECON 102 and ACE 100.   
4516      Credit is not given for ECON 102 and ACE 100.   
4517      Credit is not given for ECON 102 and ACE 100.   
4518      Credit is not given for ECON 102 and ACE 100.   
...                                                 ...   
4717  Approved for both letter and S/U grading. May ...   
4718  Approved for S/U grading only. May be repeated...   
4719  Approved for S/U grading only. May be repeated...   
4720  Approved for S/U grading only. May be repeated...   
4721  Approved for S/U grading only. May be repeated...   

                       Degree Attributes  ... Section Status  \
4514  Social & Beh Sci - Soc Sci course.  ...              A   
4515  Social & Beh Sci - Soc Sci course.  ...              A   
4516  Social & Beh Sci - Soc Sci course.  ...              A   
4517  Social & Beh Sci - Soc Sci course.  ...              A   
4518  Social & Beh Sci - Soc Sci course.  ...              A   
...                                  ...  ...            ...   
4717                                 NaN  ...              A   
4718                                 NaN  ...              A   
4719                                 NaN  ...              A   
4720                                 NaN  ...              A   
4721                                 NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
4514             Closed  Discussion/Recitation       DIS   08:00 AM  08:50 AM   
4515               Open  Discussion/Recitation       DIS   09:00 AM  09:50 AM   
4516             Closed  Discussion/Recitation       DIS   10:00 AM  10:50 AM   
4517               Open  Discussion/Recitation       DIS   11:00 AM  11:50 AM   
4518             Closed  Discussion/Recitation       DIS   12:00 PM  12:50 PM   
...                 ...                    ...       ...        ...       ...   
4717  Open (Restricted)                Seminar        M3   03:30 PM  05:20 PM   
4718  Open (Restricted)             Conference        C1   02:00 PM  03:20 PM 

### Puzzle 4.2:

Select **five random course sections** from your major, using `mydf` and pandas functions:

In [40]:

random_courses = mydf.sample(n=5)
random_courses


Year  Term YearTerm Subject  Number                           Name  \
4603  2024  Fall  2024-fa    ECON     210        Environmental Economics   
4709  2024  Fall  2024-fa    ECON     590  Individual Study and Research   
4669  2024  Fall  2024-fa    ECON     490            Topics in Economics   
4623  2024  Fall  2024-fa    ECON     303     Inter Macroeconomic Theory   
4705  2024  Fall  2024-fa    ECON     549        Environmental Economics   

                                            Description   Credit Hours  \
4603  Same as ACE 210, ENVS 210, NRES 210, and UP 21...       3 hours.   
4709  Directed reading and research. Approved for bo...  0 TO 4 hours.   
4669  Special topics in advanced economics within a ...  3 OR 4 hours.   
4623  The modern theory of the determination of the ...       3 hours.   
4705  Examines both theory and policy applications i...       4 hours.   

                                           Section Info  \
4603  Same as ACE 210, ENVS 210, NRES 210, and UP 21...   
4709  Approved for both letter and S/U grading. May ...   
4669  3 undergraduate hours. 4 graduate hours. May b...   
4623  Prerequisite: ECON 102; ECON 103; and one of M...   
4705  Same as ACE 516. Prerequisite: ECON 302 or con...   

                       Degree Attributes  ... Section Status  \
4603  Social & Beh Sci - Soc Sci course.  ...              A   
4709                                 NaN  ...              A   
4669                                 NaN  ...              A   
4623                                 NaN  ...              A   
4705                                 NaN  ...              A   

               Enrollment Status                   Type Type Code Start Time  \
4603               CrossListOpen                Lecture       LEC   02:00 PM   
4709           Open (Restricted)     Lecture-Discussion        S3   12:30 PM   
4669                      Closed     Lecture-Discussion        S3   08:00 AM   
4623                      Closed  Discussion/Recitation       DIS   11:00 AM   
4705  CrossListOpen (Restricted)     Lecture-Discussion       LCD   01:30 PM   

      End Time Days of Week Room           Building  \
4603  02:50 PM           MW  103       Mumford Hall   
4709  01:50 PM            T  119  David Kinley Hall   
4669  09:20 AM           MW  119  David Kinley Hall   
4623  11:50 AM            M  113  David Kinley Hall   
4705  02:50 PM           TR  426       Mumford Hall   

                       Instructors  
4603                      Damon, L  
4709             Lemus Encalada, J  
4669                       Song, L  
4623  Bebczuk, R;Okada Da Silva, V  
4705                   Skidmore, M  

[5 rows x 27 columns]

### Puzzle 4.3:

Select **the seven course sections with the smallest course number** from your major:

In [41]:

smallest_courses = mydf.sort_values(by='Number').head(7)
smallest_courses


Year  Term YearTerm Subject  Number                      Name  \
4514  2024  Fall  2024-fa    ECON     102  Microeconomic Principles   
4535  2024  Fall  2024-fa    ECON     102  Microeconomic Principles   
4536  2024  Fall  2024-fa    ECON     102  Microeconomic Principles   
4538  2024  Fall  2024-fa    ECON     102  Microeconomic Principles   
4539  2024  Fall  2024-fa    ECON     102  Microeconomic Principles   
4540  2024  Fall  2024-fa    ECON     102  Microeconomic Principles   
4541  2024  Fall  2024-fa    ECON     102  Microeconomic Principles   

                                            Description Credit Hours  \
4514  Introduction to the functions of individual de...     3 hours.   
4535  Introduction to the functions of individual de...     3 hours.   
4536  Introduction to the functions of individual de...     3 hours.   
4538  Introduction to the functions of individual de...     3 hours.   
4539  Introduction to the functions of individual de...     3 hours.   
4540  Introduction to the functions of individual de...     3 hours.   
4541  Introduction to the functions of individual de...     3 hours.   

                                       Section Info  \
4514  Credit is not given for ECON 102 and ACE 100.   
4535  Credit is not given for ECON 102 and ACE 100.   
4536  Credit is not given for ECON 102 and ACE 100.   
4538  Credit is not given for ECON 102 and ACE 100.   
4539  Credit is not given for ECON 102 and ACE 100.   
4540  Credit is not given for ECON 102 and ACE 100.   
4541  Credit is not given for ECON 102 and ACE 100.   

                       Degree Attributes  ... Section Status  \
4514  Social & Beh Sci - Soc Sci course.  ...              A   
4535  Social & Beh Sci - Soc Sci course.  ...              A   
4536  Social & Beh Sci - Soc Sci course.  ...              A   
4538  Social & Beh Sci - Soc Sci course.  ...              A   
4539  Social & Beh Sci - Soc Sci course.  ...              A   
4540  Social & Beh Sci - Soc Sci course.  ...              A   
4541  Social & Beh Sci - Soc Sci course.  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
4514             Closed  Discussion/Recitation       DIS   08:00 AM  08:50 AM   
4535             Closed  Discussion/Recitation       DIS   08:00 AM  08:50 AM   
4536             Closed  Discussion/Recitation       DIS   09:00 AM  09:50 AM   
4538             Closed  Discussion/Recitation       DIS   11:00 AM  11:50 AM   
4539             Closed  Discussion/Recitation       DIS   12:00 PM  12:50 PM   
4540             Closed  Discussion/Recitation       DIS   01:00 PM  01:50 PM   
4541             Closed  Discussion/Recitation       DIS   02:00 PM  02:50 PM   

     Days of Week Room           Building  \
4514            F  119  David Kinley Hall   
4535            F  432             Armory   
4536            F  432             Armory   
4538            F  432             Armory   
4539            F  432             Armory   
4540            F  432             Armory   
4541            F  432             Armory   

                                 Instructors  
4514  Ambardar, S;Cueva Chavez, R;DiIanni, I  
4535              Buckley, B;Chen, Z;Jung, Y  
4536              Buckley, B;Chen, Z;Jung, Y  
4538              Buckley, B;Chen, Z;Jung, Y  
4539                      Buckley, B;Jung, Y  
4540                      Buckley, B;Jung, Y  
4541              Buckley, B;Jung, Y;Wang, X  

[7 rows x 27 columns]

### Puzzle 4.4:

Select **the nine course sections with the highest course number** from your major:

In [42]:

highest_courses = mydf.sort_values(by='Number', ascending=False).head(9)
highest_courses


Year  Term YearTerm Subject  Number                           Name  \
4721  2024  Fall  2024-fa    ECON     598  Workshop and Research Seminar   
4720  2024  Fall  2024-fa    ECON     598  Workshop and Research Seminar   
4719  2024  Fall  2024-fa    ECON     598  Workshop and Research Seminar   
4718  2024  Fall  2024-fa    ECON     598  Workshop and Research Seminar   
4713  2024  Fall  2024-fa    ECON     590  Individual Study and Research   
4708  2024  Fall  2024-fa    ECON     590  Individual Study and Research   
4710  2024  Fall  2024-fa    ECON     590  Individual Study and Research   
4711  2024  Fall  2024-fa    ECON     590  Individual Study and Research   
4712  2024  Fall  2024-fa    ECON     590  Individual Study and Research   

                                            Description   Credit Hours  \
4721  Workshops are offered in all areas of speciali...       2 hours.   
4720  Workshops are offered in all areas of speciali...       2 hours.   
4719  Workshops are offered in all areas of speciali...       2 hours.   
4718  Workshops are offered in all areas of speciali...       2 hours.   
4713  Directed reading and research. Approved for bo...  0 TO 4 hours.   
4708  Directed reading and research. Approved for bo...  0 TO 4 hours.   
4710  Directed reading and research. Approved for bo...  0 TO 4 hours.   
4711  Directed reading and research. Approved for bo...  0 TO 4 hours.   
4712  Directed reading and research. Approved for bo...  0 TO 4 hours.   

                                           Section Info Degree Attributes  \
4721  Approved for S/U grading only. May be repeated...               NaN   
4720  Approved for S/U grading only. May be repeated...               NaN   
4719  Approved for S/U grading only. May be repeated...               NaN   
4718  Approved for S/U grading only. May be repeated...               NaN   
4713  Approved for both letter and S/U grading. May ...               NaN   
4708  Approved for both letter and S/U grading. May ...               NaN   
4710  Approved for both letter and S/U grading. May ...               NaN   
4711  Approved for both letter and S/U grading. May ...               NaN   
4712  Approved for both letter and S/U grading. May ...               NaN   

      ... Section Status  Enrollment Status                Type Type Code  \
4721  ...              A  Open (Restricted)          Conference        C4   
4720  ...              A  Open (Restricted)          Conference        C3   
4719  ...              A  Open (Restricted)          Conference        C2   
4718  ...              A  Open (Restricted)          Conference        C1   
4713  ...              A  Open (Restricted)  Lecture-Discussion        S9   
4708  ...              A  Open (Restricted)  Lecture-Discussion        S5   
4710  ...              A  Open (Restricted)  Lecture-Discussion        S2   
4711  ...              A  Open (Restricted)  Lecture-Discussion        S1   
4712  ...              A  Open (Restricted)          Conference       CNF   

     Start Time  End Time Days of Week  Room           Building  \
4721   02:00 PM  03:20 PM            F   317  David Kinley Hall   
4720   02:00 PM  03:20 PM            R   317  David Kinley Hall   
4719   02:00 PM  03:20 PM            T   317  David Kinley Hall   
4718   02:00 PM  03:20 PM           MW   317  David Kinley Hall   
4713   09:30 AM  10:50 AM           MW   226       Wohlers Hall   
4708   09:30 AM  10:50 AM           TR   106  David Kinley Hall   
4710   09:30 AM  10:50 AM           MW   138       Wohlers Hall   
4711   11:00 AM  12:20 PM           TR  215B  David Kinley Hall   
4712   12:30 PM  01:50 PM         MTWR   119  David Kinley Hall   

         Instructors  
4721         Shea, J  
4720    Durandard, T  
4719       Howard, G  
4718  Borgschulte, M  
4713         Shea, J  
4708       Deltas, G  
4710        Chung, E  
4711          Lee, J  
4712             NaN  

[9 rows x 27 columns]

### Puzzle 4.5:

Select **a random sample of 10% of the course sections** from your major:

In [43]:
sample_10_percent = mydf.sample(frac=0.1)
sample_10_percent

Year  Term YearTerm Subject  Number  \
4547  2024  Fall  2024-fa    ECON     102   
4531  2024  Fall  2024-fa    ECON     102   
4660  2024  Fall  2024-fa    ECON     483   
4527  2024  Fall  2024-fa    ECON     102   
4642  2024  Fall  2024-fa    ECON     447   
4541  2024  Fall  2024-fa    ECON     102   
4588  2024  Fall  2024-fa    ECON     202   
4606  2024  Fall  2024-fa    ECON     302   
4644  2024  Fall  2024-fa    ECON     448   
4516  2024  Fall  2024-fa    ECON     102   
4578  2024  Fall  2024-fa    ECON     199   
4567  2024  Fall  2024-fa    ECON     103   
4579  2024  Fall  2024-fa    ECON     199   
4711  2024  Fall  2024-fa    ECON     590   
4525  2024  Fall  2024-fa    ECON     102   
4695  2024  Fall  2024-fa    ECON     530   
4635  2024  Fall  2024-fa    ECON     415   
4688  2024  Fall  2024-fa    ECON     502   
4573  2024  Fall  2024-fa    ECON     103   
4521  2024  Fall  2024-fa    ECON     102   
4636  2024  Fall  2024-fa    ECON     415   

                                      Name  \
4547              Microeconomic Principles   
4531              Microeconomic Principles   
4660           Econ of Innovation and Tech   
4527              Microeconomic Principles   
4642            Economics of the Workplace   
4541              Microeconomic Principles   
4588                 Economic Statistics I   
4606            Inter Microeconomic Theory   
4644  Employee Compensation and Incentives   
4516              Microeconomic Principles   
4578            Undergraduate Open Seminar   
4567              Macroeconomic Principles   
4579            Undergraduate Open Seminar   
4711         Individual Study and Research   
4525              Microeconomic Principles   
4695                Microeconomic Theory I   
4635               Environmental Economics   
4688                   Economic Statistics   
4573              Macroeconomic Principles   
4521              Microeconomic Principles   
4636               Environmental Economics   

                                            Description   Credit Hours  \
4547  Introduction to the functions of individual de...       3 hours.   
4531  Introduction to the functions of individual de...       3 hours.   
4660  Examines the economic factors shaping innovati...  2 TO 4 hours.   
4527  Introduction to the functions of individual de...       3 hours.   
4642  Application of economic theory to the relation...  3 OR 4 hours.   
4541  Introduction to the functions of individual de...       3 hours.   
4588  Introduction of basic concepts in statistics i...       3 hours.   
4606  Microeconomic analysis including value and dis...       3 hours.   
4644  Employee compensation is a critical tool for o...  3 OR 4 hours.   
4516  Introduction to the functions of individual de...       3 hours.   
4578  Approved for both letter and S/U grading. May ...  0 TO 5 hours.   
4567  Introduction to the theory of determination of...       3 hours.   
4579  Approved for both letter and S/U grading. May ...  0 TO 5 hours.   
4711  Directed reading and research. Approved for bo...  0 TO 4 hours.   
4525  Introduction to the functions of individual de...       3 hours.   
4695  Emphasizes microeconomic theory particularly t...       4 hours.   
4635  Application of economic theory to topical issu...  3 OR 4 hours.   
4688  Classical statistics and regression analysis; ...       4 hours.   
4573  Introduction to the theory of determination of...       3 hours.   
4521  Introduction to the functions of individual de...       3 hours.   
4636  Application of economic theory to topical issu...  3 OR 4 hours.   

                                           Section Info  \
4547      Credit is not given for ECON 102 and ACE 100.   
4531      Credit is not given for ECON 102 and ACE 100.   
4660  3 undergraduate hours. 2 or 4 graduate hours. ...   
4527      Credit is not given for ECON 102 and ACE 100.   
4642  3 undergraduate hours. 4 graduate hours. Prere...   
4541      Credit is

### ✔️ Checkpoint ✔️

There are no test cases for Part 4 (since everyone has different majors), but go double check your work!  There's just one section left to go! :)

<hr style="color: #DD3403;">

# Part 5: Growth of the IS major

The IS major is still a new relatively new major here at UIUC. The final puzzle is for you to analyze how much the IS major has grown since 2019.

### Puzzle 5.1: Loading the Course Catalog Dataset from 2019

Using the name `df2019` instead of `df`, load the "Fall 2019 Course Catalog Dataset" that can be found at the following URL:
```
https://waf.cs.illinois.edu/discovery/course-catalog-fa19.csv
```

In [44]:
import pandas as pd
df2019 = pd.read_csv("https://waf.cs.illinois.edu/discovery/course-catalog-fa19.csv")
df2019

Year  Term YearTerm Subject  Number  \
0     2019  Fall  2019-fa     AAS     100   
1     2019  Fall  2019-fa     AAS     105   
2     2019  Fall  2019-fa     AAS     120   
3     2019  Fall  2019-fa     AAS     199   
4     2019  Fall  2019-fa     AAS     200   
...    ...   ...      ...     ...     ...   
8584  2019  Fall  2019-fa    ZULU     202   
8585  2019  Fall  2019-fa    ZULU     403   
8586  2019  Fall  2019-fa    ZULU     404   
8587  2019  Fall  2019-fa    ZULU     405   
8588  2019  Fall  2019-fa    ZULU     406   

                                       Name  \
0              Intro Asian American Studies   
1     Introduction to Arab American Studies   
2             Intro to Asian Am Pop Culture   
3                Undergraduate Open Seminar   
4                      U.S. Race and Empire   
...                                     ...   
8584                     Elementary Zulu II   
8585                    Intermediate Zulu I   
8586                   Intermediate Zulu II   
8587                        Advanced Zulu I   
8588                       Advanced Zulu II   

                                            Description   Credit Hours  \
0     Interdisciplinary introduction to the basic co...       3 hours.   
1     Interdisciplinary introduction to the basic co...       3 hours.   
2     Introductory understanding of the way U.S. pop...       3 hours.   
3              May be repeated to a maximum of 6 hours.  1 TO 5 hours.   
4     Invites students to examine histories and narr...       3 hours.   
...                                                 ...            ...   
8584  Continuation of ZULU 201 with introduction of ...       5 hours.   
8585  Survey of more advanced grammar; emphasis on i...       4 hours.   
8586  Continuation of ZULU 403; emphasis on increasi...       4 hours.   
8587  Third year Zulu with emphasis on conversationa...       3 hours.   
8588  Continuation of Zulu 405 with increased emphas...       3 hours.   

                                           Section Info  \
0                                                   NaN   
1                                                   NaN   
2                                                   NaN   
3                                                   NaN   
4                                      Same as LLS 200.   
...                                                 ...   
8584  Same as AFST 252. Participation in the languag...   
8585                                                NaN   
8586                                                NaN   
8587                                                NaN   
8588                                                NaN   

                                      Degree Attributes  
0     Social & Beh Sci - Soc Sci, and Cultural Studi...  
1                Cultural Studies - US Minority course.  
2                Cultural Studies - US Minority course.  
3                                                   NaN  
4     Humanities - Hist & Phil, and Cultural Studies...  
...                                                 ...  
8584                                                NaN  
8585                                                NaN  
8586                                                NaN  
8587                                                NaN  
8588                                                NaN  

[8589 rows x 10 columns]

### Puzzle 5.2: Data Science Analysis

In the following cells,
- Select only the rows from the `IS` subject from the `df2019` dataset,
- Select only the rows from the `IS` subject from the `df` dataset (this contains the Fall 2024 data),
- Complete the reflection question below (Puzzle 5.3).  Feel free to write more code to answer it.

In [45]:
df2019_IS = df2019[df2019['Subject'] == 'IS']
df2019_IS

Year  Term YearTerm Subject  Number  \
4843  2019  Fall  2019-fa      IS     107   
4844  2019  Fall  2019-fa      IS     199   
4845  2019  Fall  2019-fa      IS     202   
4846  2019  Fall  2019-fa      IS     310   
4847  2019  Fall  2019-fa      IS     351   
...    ...   ...      ...     ...     ...   
4930  2019  Fall  2019-fa      IS     591   
4931  2019  Fall  2019-fa      IS     592   
4932  2019  Fall  2019-fa      IS     593   
4933  2019  Fall  2019-fa      IS     594   
4934  2019  Fall  2019-fa      IS     599   

                                             Name  \
4843                       Data Science Discovery   
4844                   Undergraduate Open Seminar   
4845                     Social Aspects Info Tech   
4846                  Computing in the Humanities   
4847  The Design of Usable Information Interfaces   
...                                           ...   
4930                                    Practicum   
4931                            Independent Study   
4932                                  CAS Project   
4933                Information Sciences Practice   
4934                              Thesis Research   

                                            Description    Credit Hours  \
4843          Same as CS 107 and STAT 107. See STAT 107        4 hours.   
4844       Undergraduate Open Seminar. May be repeated.   1 TO 5 hours.   
4845       Same as INFO 202 and MACS 202. See INFO 202.        3 hours.   
4846  Explores use and application of technology to ...        3 hours.   
4847  Examines issues of Human Computer Interaction ...        3 hours.   
...                                                 ...             ...   
4930  Supervised field experience of professional-le...        2 hours.   
4931  Permits the intermediate or advanced student o...   2 TO 4 hours.   
4932  Individual study of a problem in library and i...   0 TO 8 hours.   
4933  Full-time or part-time practice of of any info...        0 hours.   
4934  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info  \
4843          Same as CS 107 and STAT 107. See STAT 107   
4844                                                NaN   
4845       Same as INFO 202 and MACS 202. See INFO 202.   
4846                                                NaN   
4847                  Prerequisite: Sophomore standing.   
...                                                 ...   
4930  2 graduate hours. No professional credit. Appr...   
4931  2 to 4 graduate hours. No professional credit....   
4932  0 to 8 graduate hours. No professional credit....   
4933  0 graduate hours. No professional credit. Appr...   
4934  0 to 16 graduate hours. No professional credit...   

                       Degree Attributes  
4843    Quantitative Reasoning I course.  
4844                                 NaN  
4845  Social & Beh Sci - Soc Sci course.  
4846                                 NaN  
4847                                 NaN  
...                                  ...  
4930                                 NaN  
4931                                 NaN  
4932                                 NaN  
4933                                 NaN  
4934                                 NaN  

[92 rows x 10 columns]

In [46]:
df_IS = df[df['Subject'] == 'IS']
df_IS

Year  Term YearTerm Subject  Number  \
7003  2024  Fall  2024-fa      IS     100   
7004  2024  Fall  2024-fa      IS     100   
7005  2024  Fall  2024-fa      IS     100   
7006  2024  Fall  2024-fa      IS     100   
7007  2024  Fall  2024-fa      IS     100   
...    ...   ...      ...     ...     ...   
7303  2024  Fall  2024-fa      IS     599   
7304  2024  Fall  2024-fa      IS     599   
7305  2024  Fall  2024-fa      IS     599   
7306  2024  Fall  2024-fa      IS     599   
7307  2024  Fall  2024-fa      IS     599   

                                                  Name  \
7003  Exploring the iSchool with a Human-Centered Lens   
7004  Exploring the iSchool with a Human-Centered Lens   
7005  Exploring the iSchool with a Human-Centered Lens   
7006  Exploring the iSchool with a Human-Centered Lens   
7007  Exploring the iSchool with a Human-Centered Lens   
...                                                ...   
7303                                   Thesis Research   
7304                                   Thesis Research   
7305                                   Thesis Research   
7306                                   Thesis Research   
7307                                   Thesis Research   

                                            Description    Credit Hours  \
7003  This course introduces students to the School ...        1 hours.   
7004  This course introduces students to the School ...        1 hours.   
7005  This course introduces students to the School ...        1 hours.   
7006  This course introduces students to the School ...        1 hours.   
7007  This course introduces students to the School ...        1 hours.   
...                                                 ...             ...   
7303  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   
7304  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   
7305  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   
7306  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   
7307  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info Degree Attributes  \
7003  Prerequisite: Restricted to Majors Only; First...               NaN   
7004  Prerequisite: Restricted to Majors Only; First...               NaN   
7005  Prerequisite: Restricted to Majors Only; First...               NaN   
7006  Prerequisite: Restricted to Majors Only; First...               NaN   
7007  Prerequisite: Restricted to Majors Only; First...               NaN   
...                                                 ...               ...   
7303  0 to 16 graduate hours. No professional credit...               NaN   
7304  0 to 16 graduate hours. No professional credit...               NaN   
7305  0 to 16 graduate hours. No professional credit...               NaN   
7306  0 to 16 graduate hours. No professional credit...               NaN   
7307  0 to 16 graduate hours. No professional credit...               NaN   

      ... Section Status  Enrollment Status                Type Type Code  \
7003  ...              A             Closed  Lecture-Discussion       LCD   
7004  ...              A             Closed  Lecture-Discussion       LCD   
7005  ...              A             Closed  Lecture-Discussion       LCD   
7006  ...              A             Closed  Lecture-Discussion       LCD   
7007  ...              A             Closed  Lecture-Discussion       LCD   
...   ...            ...                ...                 ...       ...   
7303  ...              A               Open   Independent Study       IND   
7304  ...              A               Open   Independent Study       IND   
7305  ...              A               Open   Independent Study       IND   
7306  ...              A               Open   Independent Study       IND   
7307  ...              A               Open   Independent Study       IND   

     Start Time  

In [47]:
# Feel free to add more code cells if needed by hovering over the area below the cell and pressing "+ Code"!

### Puzzle 5.3: Reflection on the Growth of Information Science

In the follow cell, write **at least one complete sentence explaining what you found** related to the growth of the IS program.  In your answer, **use actual data** (ex: by how much did it grow?)

*(In 2019, the IS program offered 92 courses, while in 2024, the number of courses increased to 305. )*

<hr style="color: #DD3403;">

# Submission

You're almost done!  All you need to do is to commit your lab to GitHub:

1.  ⚠️ **Make certain to save your work.** ⚠️ To do this, go to **File => Save All**

2.  After you have saved, exit this notebook and follow the Canvas instructions to commit this lab to your Git repository!

3. Your TA will grade your submission and provide you feedback after the lab is due. :)